support: add SplitView for parsing strings with delimiters

tylershunt opened the pull request:
I needed one of these for some string parsing. Couldn't find one in
fmt, and it seems like such a small thing to take a new dependency on.

Interface intended to be very close to python's `string.split()`

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

roshandathathri commented on libcusp/include/galois/graphs/DistributedGraph.h:
```diff
@@ -577,11 +589,20 @@ class DistGraph {
+  auto&
+  getDataIndex(GraphNode N,
+               galois::MethodFlag mflag = galois::MethodFlag::UNPROTECTED) {
+    return std::get<Index>(graph.getData(N, mflag));
```
getData() is not defined if the node type is a tuple, so how does this work? 

I don't understand the goal here. The node type has been changed from structure to tuple if I understand right. Instead of an array of structures, the graph has an array of tuples. However, to support property graphs, we would need the graph to provide access to a structure/tuple of arrays. What am I missing here?

ddn0 replied:
> After this lands, next steps are: replacing the std::tuple node data with direct access to underlying property file graph buffers.

This particular PR just refactors things in advance of using property file graph buffers directly while keeping test coverage and the same functionality as before. So we can be doubly sure that the large change of property graphs doesn't break any applications.

Once this is in, graph operations will use opaque identifiers to access fields rather than going through a struct. And the subsequent changes (like modifying DistGraph) can be done without changing application code.

roshandathathri replied:
Thanks, the goal makes sense to me now.

What about this: getData() is not defined if the node type is a tuple, so how does this work?

l-hoang replied:
Might be wrong on this, but I'm pretty sure getData isn't used anywhere anymore with the DistGraph.

roshandathathri replied:
My point is that it's being used in getDataIndex!

roshandathathri replied:
Never mind, "graph" here is not DistGraph, so it works.

ddn0 replied:
Yup, `getData` in its current form is no longer used. I think now that everything has switched over to `getDataIndex`, we probably should reclaim `getData` solely for indexed access (and remove `getDataIndex`). I didn't do this in this PR because I didn't want to deal with issues if there were concurrent changes in flight (and it is easier update things when both forms can coexist for a while).

roshandathathri replied:
Agreed.

l-hoang commented:
1) My guess is that the performance hit would be minimal, but of course that needs testing.
3) There will still be a performance hit for applications that rely on array of structs (MRBC is the app that sticks out the most for this; performance was heavily dependent on locality from when we were testing this a few years ago).

ddn0 commented:
> (1) I'm not sure if index is better as a template parameter or a function parameter. I'm leaning towards the latter. If we want the user to specify the property/index, then the code will become messy if it is a template parameter. I don't think it would be a performance hit if we make it a function parameter. What do you think? We should decide this in this PR.

Happy to hear feedback on the interface!

The point of the template parameter is not for performance but compile-time checking. Although this is not very useful with bare enums, the nice-to-have improvement is to use types/`enum class` as keys rather than integers. Then, we would be able to check if someone used the wrong field for a graph. This interface could be layered on top of a PropertyFileGraph to give a typed view. I was planning on calling this the `PropertyGraph` (the PR description gives a flavor of what it would be like).

It is not strictly necessary for now, so that's why I left indices as pure integers but if we want to go this path in the future, it would mean accesses through the typed view would access fields with a template parameter.

Two open questions I had were:
- Whether there needs to be an untyped class that sits between PropertyFileGraph and "PropertyGraph" or if all un-typed accesses can be done via PropertyFileGraph.
- Whether internal functions (SyncStructures, etc.) use PropertyFileGraph (or buffers) directly or should they go through the client facing PropertyGraph. I think we should do the former here, which might address some of @roshandathathri 's concerns about messiness.

Commit:
support: add SplitView for parsing strings with delimiters

I needed one of these for some string parsing. Couldn't find one in
fmt, and it seems like such a small thing to take a new dependency on.

Interface intended to be very close to python's `string.split()`

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

witchel commented on libsupport/include/katana/Strings.h:
```diff
@@ -33,6 +34,13 @@ KATANA_EXPORT std::string TrimSuffix(
+/// SplitView returns a list of words in \param s using \param sep as the
+/// delimiter string. Splits at most \param max times (so there will be at most
+/// max + 1 entries in the output)
+KATANA_EXPORT std::vector<std::string_view> SplitView(
```
Nice comment.

ddn0 replied:
I don't think doxygen `param` supports inline formatting

https://www.doxygen.nl/manual/commands.html#cmdparam

so I think this will render as


```
SplitView returns a list of words in.

s | using
sep | as the delimiter string. Splits at most
max | times (so there will be at most max + 1 entries in the output)
```

tylershunt replied:
Oh. Woops! i've been using this style for months. Should have looked at the docs.

witchel reviewed (approved):
Looks good.  Pretty weird it isn't in the standard library.
